File: Problem_Sets/LeetCode/BalancedBinaryTreee/BalancedBinaryTree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        def getHeight(root: Optional[TreeNode]) -> int:
            if not root:
                return 0
            lHeight = getHeight(root.left)
            rHeight = getHeight(root.right)
            if lHeight == -1 or rHeight == -1 or abs(lHeight - rHeight) > 1:
                return -1
            return max(lHeight, rHeight) + 1
        
        if getHeight(root) < 0:
            return False
        return True
        

File: Problem_Sets/LeetCode/BalancedBinaryTreee/BalancedBinaryTree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int getHeight(TreeNode* root) {
        if (root == nullptr) {
            return 0;
        }

        // Recursively check left and right subtrees
        int leftHeight = getHeight(root->left);
        int rightHeight = getHeight(root->right);

        // Check if subtrees are balanced and height difference is less than or equal to 1
        if (leftHeight == -1 || rightHeight == -1 || abs(leftHeight - rightHeight) > 1) {
            return -1; // Unbalanced tree
        }

        // Return height of the current node
        return max(leftHeight, rightHeight) + 1;
    }

    bool isBalanced(TreeNode* root) {
        if (getHeight(root) < 0) {
            return false;
        }
        return true;
    }
};